x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/GrantApiKeyRequest.java

```diff
@@ -86,6 +89,21 @@ public void setPassword(SecureString password) {
         public void setAccessToken(SecureString accessToken) {
             this.accessToken = accessToken;
         }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject()
+                    .field("type", type)
+                    .field("username", username);
+            if (params.paramAsBoolean(AuditToXContentParams.INCLUDE_CREDENTIALS, false)) {
+                builder.field("password", password != null ? password.toString() : null);
+                builder.field("access_token", accessToken != null ? accessToken.toString() : null);
+            } else {
+                builder.field("password", password != null ? "<redacted>" : null);
```

Comment:
If we are redacting, should we just always show `<redacted>` regardless of whether the value is null or not? It is better from the information disclosure perspective.

Reply:
It is a good question, but I don't think we have to be that cautious, because the existence of a password doesn't give away much information. I think the audit log overall gives away much more details on the activity on the system.

IMO, the `null` vs `<redacted>` thing is not even something that we should document and test. I'm comfortable with only saying that "credentials and credential hashes are not printed in the audit log".

Reply:
> IMO, the null vs <redacted> thing is not even something that we should document and test. I'm comfortable with only saying that "credentials and credential hashes are not printed in the audit log".

Err, except that the put user action is an example where `null` vs `<redacted>` IS significant. if the `password_hash` field is `<redacted>` then the put user action updated the password, whereas if it is `null` then user password remains unchanged.

Therefore, the password/password_hash field will be `null` if and only if the field is not present in the request body. Otherwise it will be `<redacted>`. I'lll make sure to test for it.